File: python/src/runapi/suno/resources/convert_audio.py
"""Suno convert_audio resource."""

from __future__ import annotations

from typing import Any, Dict

from runapi.core import Resource

from .. import _validators
from ..types import CompletedConvertAudioResponse, ConvertAudioResponse


class ConvertAudio(Resource):
    """Convert a track to a WAV file."""

    ENDPOINT = "/api/v1/suno/convert_audio"

    RESPONSE_CLASS = ConvertAudioResponse
    COMPLETED_RESPONSE_CLASS = CompletedConvertAudioResponse

    def run(self, **params: Any) -> Any:
        """Convert a track to WAV and poll until it completes.

        Args:
            **params: convert-audio parameters.

        Returns:
            The completed (narrowed) response.
        """
        task = self.create(**params)
        return self._poll_until_complete(lambda: self.get(task.id))

    def create(self, **params: Any) -> Any:
        """Create a convert-audio task and return immediately with an id.

        Args:
            **params: convert-audio parameters.

        Returns:
            The task creation result with an id.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def get(self, id: str) -> Any:
        """Fetch the current status of a convert-audio task.

        Args:
            id: The task id returned by ``create``.

        Returns:
            The current task status.
        """
        return self._request("get", f"{self.ENDPOINT}/{id}")

    def _validate_params(self, params: Dict[str, Any]) -> None:
        _validators.validate_convert_audio(params)

File: ruby/lib/runapi/suno/resources/convert_audio.rb
# frozen_string_literal: true

module RunApi
  module Suno
    module Resources
      # Converts a generated track to WAV format.
      class ConvertAudio
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/suno/convert_audio"
        RESPONSE_CLASS = Types::ConvertAudioResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedConvertAudioResponse

        def initialize(http)
          @http = http
        end

        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          Validators.validate_convert_audio!(params, self)
        end
      end
    end
  end
end
